### DISCUSSION TITLE
Multiprocess slow start for ops

### DISCUSSION THREAD

**Discussion by joshuataylor** [dagster-io/dagster#7338] [Q&A] [answered]
Hi All

With dagster in production we're seeing ops take 5+ seconds to start, even if nothing else is running at the same time.
![image](https://user-images.githubusercontent.com/225131/162109152-15f8bf7f-d672-475c-a20e-3234872e97a2.png)

We're using:
Python 3.8.12 - just upgraded to 3.10.4 and it has the same issue
Dagster: 0.14.7
Backing store: Postgres (AWS RDS)
Host: Ubuntu on AWS EC2 Graviton2 (but also has issues on AMD64) `Linux ip-xxx 5.13.0-1017-aws #19~20.04.1-Ubuntu SMP Mon Mar 7 12:55:31 UTC 2022 aarch64 aarch64 aarch64 GNU/Linux)` Ubuntu 20.04.4 LTS

**Comment by joshuataylor** [comment]
We have found that solids don't seem to have this, thinking it could be a subprocess thing?
![image](https://user-images.githubusercontent.com/225131/164128157-524e200c-2ce9-4370-8ac1-eab790530f68.png)

**Reply by prha** [reply]
Yeah, this could be the overhead of spawning a new process for each op.   It may be the number of Python imports that have to happen in the process.  When comparing against solids, do you see the same latency when using the `in_process_executor`?  You can invoke this by assigning the executor on the job itself:

```
from dagster import job, op, in_process_executor

@job(executor_def=in_process_executor)
def my_job():
    ...
```

**Comment by joshuataylor** [comment]
Yeah, with `in_process_executor` I see it being pretty much instant, same as the previous way.

We do have a semi-complicated setup where we build a list of jobs dynamically from YAML, so if I had to guess this would be the reason for the delay -- does dagster bootstrap the repository list each time?

**Reply by alangenfeld** [reply]
> does dagster bootstrap the repository list each time?

Yea, when we need to load an artifact in a new subprocess we do it based on how it was loaded originally.

**Comment by alangenfeld** [comment, marked as answer]
One tool to avoid repeating the cost is to use the `start_method` `forkserver` which is available as config on the multiprocess executor. The first op will still pay the init cost to create the template process, but each subsequent op should be faster as it forks that template process instead of starting from scratch. You may need to explicitly set `preload_modules` in the config if the default behavior doesn't load the necessary modules. 

context: https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods

**Comment by isaac-jordan** [comment]
Tried this myself and I think I *maybe* see a small increase using forkserver and a list of preloaded modules, but it's still taking approx 8 seconds to start any op (`Launching subprocess` to `Starting initialization of resources...`).

**Reply by martincpt** [reply]
Unfortunately, I agree. We lose more time waiting for processes to launch than we gain with fanout and multiprocessing. We've migrated our pipelines back to using in_process_executor and now execute our operations in asynchronous batches with a custom retry mechanism. Performance immediately improved. The run with multiprocessing and fanout took 40–60 minutes; now it takes 1–2 minutes tops.

**Reply by joshuataylor** [reply]
Dang, that is a massive difference, and would be a great casestudy/usecase if you could replicate a minimal reproducible example for it.
